arch/xtensa/src/esp32/esp32_spi.c

```diff
@@ -0,0 +1,1010 @@
+ * Name: esp32_spi_set_regbits
+ *
+ * Description:
+ *   Set the bits of the SPI register at offset
+ *
+ * Input Parameters:
+ *   priv   - Private SPI device structure
+ *   offset - Offset to the register of interest
+ *   bits   - Bits to be set
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static inline void esp32_spi_set_regbits(struct esp32_spi_priv_s *priv,
+                                         int offset,
+                                         uint32_t bits)
+{
+  uint32_t tmp = getreg32(priv->config->reg_base + offset);
+
+  putreg32(tmp | bits, priv->config->reg_base + offset);
+}
+
+/****************************************************************************
+ * Name: esp32_spi_reset_regbits
+ *
+ * Description:
+ *   Clear the bits of the SPI register at offset
+ *
+ * Input Parameters:
+ *   priv   - Private SPI device structure
+ *   offset - Offset to the register of interest
+ *   bits   - Bits to be cleared
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static inline void esp32_spi_reset_regbits(struct esp32_spi_priv_s *priv,
+                                           int offset,
+                                           uint32_t bits)
+{
+  uint32_t tmp = getreg32(priv->config->reg_base + offset);
+
+  putreg32(tmp & (~bits), priv->config->reg_base + offset);
+}
+
+/****************************************************************************
+ * Name: esp32_spi_lock
+ *
+ * Description:
+ *   Lock or unlock the SPI device
+ *
+ * Input Parameters:
+ *   priv   - Private SPI device structure
+ *   lock   - true: Lock spi bus, false: unlock SPI bus
+ *
+ * Returned Value:
+ *   The result of lock or unlock the SPI device
+ *
+ ****************************************************************************/
+
+static int esp32_spi_lock(FAR struct spi_dev_s *dev, bool lock)
+{
+  int ret;
+  FAR struct esp32_spi_priv_s *priv = (FAR struct esp32_spi_priv_s *)dev;
+
+  if (lock)
+    {
+      ret = nxsem_wait_uninterruptible(&priv->exclsem);
+    }
+  else
+    {
+      ret = nxsem_post(&priv->exclsem);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: esp32_spi_setfrequency
+ *
+ * Description:
+ *   Set the SPI frequency.
+ *
+ * Input Parameters:
+ *   dev -       Device-specific state data
+ *   frequency - The SPI frequency requested
+ *
+ * Returned Value:
+ *   Returns the actual frequency selected
+ *
+ ****************************************************************************/
+
+static uint32_t esp32_spi_setfrequency(FAR struct spi_dev_s *dev,
+                                       uint32_t frequency)
+{
+  uint32_t reg_val;
+  FAR struct esp32_spi_priv_s *priv = (FAR struct esp32_spi_priv_s *)dev;
+  const uint32_t duty_cycle = 128;
+
+  if (frequency > ((APB_CLK_FREQ / 4) * 3))
```

Comment:
Higher level drivers use setfrequency/setmode/setbits before every SPI transaction to make sure that the bus is in the state that the driver wants it to be.  To avoid having to do all of what follows, usually we test if the frequency hasn't changed, if so we just return.  Something like:
``` C
if (priv->frequency == frequency)
  {
    /* We are already at this frequency.  Return the actual. */

    return priv->actual;
  }
```

Reply:
Good point Abdelatif! Thank you!